sklearn/feature_selection/from_model.py

```diff
@@ -123,25 +129,52 @@ class SelectFromModel(BaseEstimator, SelectorMixin, MetaEstimatorMixin):
     threshold_ : float
         The threshold value used for feature selection.
     """
-    def __init__(self, estimator, threshold=None, prefit=False, norm_order=1):
+    def __init__(self, estimator, threshold=None, prefit=False,
+                 max_features=None, norm_order=1):
         self.estimator = estimator
         self.threshold = threshold
         self.prefit = prefit
+        self.max_features = max_features
         self.norm_order = norm_order
 
+    def _check_max_features(self, X, max_features):
+        if self.max_features is None or self.max_features == 'all':
+            return
+
+        if isinstance(self.max_features, numbers.Integral):
+            if 0 <= self.max_features <= X.shape[1]:
+                return
+
+        raise ValueError(
+            "max_features should be >=0, <= n_features or 'all';"
+            " got %r. Use max_features='all' to return all features."
+            % self.max_features)
+
+    def _check_params(self, X, y):
+        X, y = check_X_y(X, y)
+        self._check_max_features(X, self.max_features)
```

Comment:
We tend to avoid such nesting and would rather have the check_max_features logic inline here.